g2p/mappings/__init__.py

```diff
@@ -176,34 +178,27 @@ def process_model_specs(self):  # noqa: C901
             self.rules = sorted(
                 # Temporarily normalize to NFD for heuristic sorting of NFC-defined rules
                 self.rules,
-                key=lambda x: (
-                    len(normalize(strip_index_notation(x.rule_input), "NFD"))
-                    if isinstance(x, Rule)
-                    else len(normalize(x["in"], "NFD"))
-                ),
+                key=lambda x: len(normalize(strip_index_notation(x.rule_input), "NFD")),
                 reverse=True,
             )
 
+        def apply_to_attributes(rule: Rule, func: Callable, *attrs):
+            for k in attrs:
+                value = getattr(rule, k)
+                if value:  # won't be None since default is ""
+                    setattr(rule, k, func(value))
+
         non_empty_mappings: List[Rule] = []
         for i, rule in enumerate(self.rules):
-            if isinstance(rule, dict):
-                rule = Rule(**rule)
             # Expand Abbreviations
-            if (
-                self.abbreviations
-                and self.rules
-                and "match_pattern" not in self.rules[0]
```

Comment:
I really don't know what this was intended to do.  Why only check `self.rules[0]`?  Why not the other rules?

Reply:
Therefore I removed this check, it seems meaningless - if the intent was not to overwrite a pre-specified `match_pattern` then this is definitely the wrong place to do that.

Reply:
I like most of this PR, but I'll have to come back and analyse this change more carefully, because I'm not clear on the original purpose of this check and therefore I'm not confident it's unnecessary. When we wrote it, there must have been a reason, I'd like to dig and find it.

Reply:
yes, I admit I didn't look at `git blame` here, it might reveal the original purpose.  not overwriting an explicit `match_pattern` is something we might want to do, for instance...

Reply:
This is the commit that added the check for "match_pattern" in the first rule: https://github.com/roedoejet/g2p/commit/8ec4f271b31f21b774619aad418266f6ee85264b @roedoejet do you remember what the intent was?

Reply:
Yeah, I think it is actually impossible for this to happen now, unless `process_model_specs`, which is called during construction of a `Mapping`, gets called explicitly later on.  Probably what we want to actually do is ensure that nobody does that.

Reply:
Ok - the way that makes the most sense to do this is to just declare that seeing `match_pattern` or `intermediate_form` in a rule (any rule) is a programmer error, since they are excluded from serialization, and if you specify them, they will get overwritten anyway, so you shouldn't do that.  Added a commit for that!